File: CodeQualityCheckerPython/фИНАЛ ПП/CodeQualityChecker/CodeQualityChecker/Services/GitHubService.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Newtonsoft.Json.Linq;

public class GitHubService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _config;

    public GitHubService(HttpClient httpClient, IConfiguration config)
    {
        _httpClient = httpClient;
        _config = config;
    }

    public (string owner, string repo, string branch, string path) ParseGitHubUrl(string url)
    {
        url = url.TrimEnd('/');
        var uri = new Uri(url);
        var segments = uri.Segments.Skip(1).ToArray();

        if (segments.Length < 2)
            throw new ArgumentException("Неверный формат URL GitHub репозитория");

        var owner = segments[0].TrimEnd('/');
        var repo = segments[1].TrimEnd('/').Replace(".git", "");
        var branch = "";
        var path = "";

        var pathIndex = -1;
        for (int i = 2; i < segments.Length; i++)
        {
            var segment = segments[i].TrimEnd('/');
            if (segment == "tree")
            {
                pathIndex = i + 1;
                if (pathIndex < segments.Length)
                    branch = segments[pathIndex].TrimEnd('/');
                break;
            }
            else if (segment == "blob")
            {
                pathIndex = i + 2;
                if (pathIndex < segments.Length)
                    branch = segments[pathIndex - 1].TrimEnd('/');
                break;
            }
        }

        if (pathIndex >= 0 && pathIndex < segments.Length)
        {
            var pathSegments = segments.Skip(pathIndex).Select(s => s.TrimEnd('/'));
            path = string.Join("/", pathSegments);
        }

        return (owner, repo, branch, path);
    }

    public async Task<string> GetDefaultBranchAsync(string owner, string repo, string? token = null)
    {
        var request = new HttpRequestMessage(HttpMethod.Get,
            $"https://api.github.com/repos/{owner}/{repo}");
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        
        if (!string.IsNullOrEmpty(token))
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        return json["default_branch"]?.ToString() ?? "main";
    }

    public async Task<string> GetFileContentAsync(string owner, string repo, string path, string branch, string? token = null)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, 
            $"https://api.github.com/repos/{owner}/{repo}/contents/{path}?ref={branch}");
        
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        
        if (!string.IsNullOrEmpty(token))
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        var content = json["content"]?.ToString();
        
        if (string.IsNullOrEmpty(content))
            throw new Exception("Не удалось получить содержимое файла");

        var decoded = Encoding.UTF8.GetString(Convert.FromBase64String(content.Replace("\n", "")));
        return decoded;
    }

    public async Task<List<GitFileInfo>> GetDirectoryContentsAsync(string owner, string repo, string path, string branch, string? token = null)
    {
        var url = string.IsNullOrEmpty(path)
            ? $"https://api.github.com/repos/{owner}/{repo}/contents?ref={branch}"
            : $"https://api.github.com/repos/{owner}/{repo}/contents/{path}?ref={branch}";

        var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        
        if (!string.IsNullOrEmpty(token))
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JArray.Parse(await response.Content.ReadAsStringAsync());
        var files = new List<GitFileInfo>();

        foreach (var item in json)
        {
            files.Add(new GitFileInfo
            {
                Name = item["name"]?.ToString() ?? "",
                Path = item["path"]?.ToString() ?? "",
                Type = item["type"]?.ToString() ?? "file",
                Size = item["size"]?.Value<long>() ?? 0,
                DownloadUrl = item["download_url"]?.ToString()
            });
        }

        return files;
    }

    public async Task<GitCommitResult> CreateCommitWithFixesAsync(
        string owner, string repo, string branch,
        string filePath, string originalContent, string fixedContent,
        string commitMessage, string prDescription, string? token = null)
    {
        if (string.IsNullOrEmpty(token))
            throw new ArgumentException("Требуется токен для создания коммита");

        var branchInfo = await GetBranchInfoAsync(owner, repo, branch, token);
        var sha = branchInfo["commit"]?["sha"]?.ToString();

        if (string.IsNullOrEmpty(sha))
            throw new Exception("Не удалось получить информацию о ветке");

        var featureBranch = $"code-quality-fix-{DateTime.Now:yyyyMMddHHmmss}";
        await CreateBranchAsync(owner, repo, featureBranch, sha, token);
        await UpdateFileAsync(owner, repo, filePath, fixedContent, commitMessage, featureBranch, token);

        var prNumber = await CreatePullRequestAsync(owner, repo, featureBranch, branch, commitMessage, prDescription, token);

        return new GitCommitResult
        {
            Success = true,
            BranchName = featureBranch,
            PullRequestUrl = $"https://github.com/{owner}/{repo}/pull/{prNumber}",
            PullRequestNumber = prNumber
        };
    }

    public async Task<bool> AddPullRequestCommentAsync(string owner, string repo, int prNumber, string comment, string? token = null)
    {
        if (string.IsNullOrEmpty(token))
            throw new ArgumentException("Требуется токен");

        var request = new HttpRequestMessage(HttpMethod.Post,
            $"https://api.github.com/repos/{owner}/{repo}/issues/{prNumber}/comments");
        
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Content = new StringContent(
            JsonSerializer.Serialize(new { body = comment }),
            Encoding.UTF8, "application/json");

        var response = await _httpClient.SendAsync(request);
        return response.IsSuccessStatusCode;
    }

    private async Task<JObject> GetBranchInfoAsync(string owner, string repo, string branch, string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Get,
            $"https://api.github.com/repos/{owner}/{repo}/branches/{branch}");
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        return JObject.Parse(await response.Content.ReadAsStringAsync());
    }

    private async Task CreateBranchAsync(string owner, string repo, string branchName, string sha, string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Post,
            $"https://api.github.com/repos/{owner}/{repo}/git/refs");
        
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Content = new StringContent(
            JsonSerializer.Serialize(new { @ref = $"refs/heads/{branchName}", sha = sha }),
            Encoding.UTF8, "application/json");

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
    }

    private async Task UpdateFileAsync(string owner, string repo, string path, string newContent, string message, string branch, string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Put,
            $"https://api.github.com/repos/{owner}/{repo}/contents/{path}");
        
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        
        var body = new
        {
            message = message,
            content = Convert.ToBase64String(Encoding.UTF8.GetBytes(newContent)),
            branch = branch
        };

        request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
    }

    private async Task<int> CreatePullRequestAsync(string owner, string repo, string head, string base_, string title, string body, string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Post,
            $"https://api.github.com/repos/{owner}/{repo}/pulls");
        
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Content = new StringContent(
            JsonSerializer.Serialize(new { head, base_, title, body }),
            Encoding.UTF8, "application/json");

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        return json["number"]?.Value<int>() ?? 0;
    }

    public async Task<bool> ValidateTokenAsync(string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, "https://api.github.com/user");
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        return response.IsSuccessStatusCode;
    }

    public async Task<string> GetUserNameAsync(string token)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, "https://api.github.com/user");
        request.Headers.UserAgent.Add(new ProductInfoHeaderValue("CodeQualityChecker", "1.0"));
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        return json["login"]?.ToString() ?? "unknown";
    }
}

public class GitFileInfo
{
    public string Name { get; set; } = "";
    public string Path { get; set; } = "";
    public string Type { get; set; } = "file";
    public long Size { get; set; }
    public string? DownloadUrl { get; set; }
}

public class GitCommitResult
{
    public bool Success { get; set; }
    public string? ErrorMessage { get; set; }
    public string? BranchName { get; set; }
    public string? PullRequestUrl { get; set; }
    public int PullRequestNumber { get; set; }
}

File: CodeQualityCheckerPython/services/github_service.py
"""
Сервис для работы с GitHub API.
"""
import base64
from datetime import datetime
from typing import List, Optional, Tuple
from urllib.parse import urlparse

import requests

from models import GitCommitResult, GitFileInfo, TokenValidationResult


class GitHubService:
    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update(
            {
                "User-Agent": "CodeQualityChecker/1.0",
                "Accept": "application/vnd.github.v3+json",
            }
        )

    def parse_github_url(self, url: str) -> Tuple[str, str, str, str]:
        url = url.strip().rstrip("/")
        parsed = urlparse(url)
        segments = [s for s in parsed.path.split("/") if s]
        if len(segments) < 2:
            raise ValueError("Неверный формат URL GitHub репозитория")
        owner = segments[0]
        repo = segments[1].replace(".git", "")
        branch = ""
        path = ""
        for i, seg in enumerate(segments[2:], start=2):
            if seg in {"tree", "blob"}:
                branch = segments[i + 1] if i + 1 < len(segments) else ""
                path = "/".join(segments[i + 2 :]) if i + 2 < len(segments) else ""
                break
        return owner, repo, branch, path

    async def get_default_branch(
        self, owner: str, repo: str, token: Optional[str] = None
    ) -> str:
        headers = dict(self.session.headers)
        if token:
            headers["Authorization"] = f"Bearer {token}"
        response = self.session.get(f"https://api.github.com/repos/{owner}/{repo}", headers=headers)
        response.raise_for_status()
        return response.json().get("default_branch", "main")

    async def get_file_content(
        self, owner: str, repo: str, path: str, branch: str, token: Optional[str] = None
    ) -> str:
        headers = dict(self.session.headers)
        if token:
            headers["Authorization"] = f"Bearer {token}"
        url = f"https://api.github.com/repos/{owner}/{repo}/contents/{path}?ref={branch}"
        response = self.session.get(url, headers=headers)
        response.raise_for_status()
        content = response.json().get("content", "")
        if not content:
            raise ValueError("Не удалось получить содержимое файла")
        return base64.b64decode(content.replace("\n", "")).decode("utf-8", errors="replace")

    async def get_directory_contents(
        self, owner: str, repo: str, path: str, branch: str, token: Optional[str] = None
    ) -> List[GitFileInfo]:
        headers = dict(self.session.headers)
        if token:
            headers["Authorization"] = f"Bearer {token}"
        url = (
            f"https://api.github.com/repos/{owner}/{repo}/contents/{path}?ref={branch}"
            if path
            else f"https://api.github.com/repos/{owner}/{repo}/contents?ref={branch}"
        )
        response = self.session.get(url, headers=headers)
        response.raise_for_status()
        data = response.json()
        if isinstance(data, dict):
            data = [data]
        return [
            GitFileInfo(
                name=item.get("name", ""),
                path=item.get("path", ""),
                type="dir" if item.get("type") == "dir" else "file",
                size=item.get("size", 0),
                download_url=item.get("download_url"),
            )
            for item in data
        ]

    async def create_commit_with_fixes(
        self,
        owner: str,
        repo: str,
        branch: str,
        file_path: str,
        fixed_content: str,
        commit_message: str,
        pr_description: str,
        token: str,
    ) -> GitCommitResult:
        if not token:
            return GitCommitResult(success=False, error_message="Требуется токен")
        headers = {**self.session.headers, "Authorization": f"Bearer {token}"}
        try:
            branch_url = f"https://api.github.com/repos/{owner}/{repo}/branches/{branch}"
            response = self.session.get(branch_url, headers=headers)
            response.raise_for_status()
            sha = response.json()["commit"]["sha"]

            feature_branch = f"code-quality-fix-{datetime.now().strftime('%Y%m%d%H%M%S')}"
            ref_data = {"ref": f"refs/heads/{feature_branch}", "sha": sha}
            response = self.session.post(
                f"https://api.github.com/repos/{owner}/{repo}/git/refs",
                headers=headers,
                json=ref_data,
            )
            response.raise_for_status()

            get_file_url = f"https://api.github.com/repos/{owner}/{repo}/contents/{file_path}?ref={branch}"
            response = self.session.get(get_file_url, headers=headers)
            response.raise_for_status()
            current_sha = response.json().get("sha")

            content_data = {
                "message": commit_message,
                "content": base64.b64encode(fixed_content.encode("utf-8")).decode("utf-8"),
                "branch": feature_branch,
            }
            if current_sha:
                content_data["sha"] = current_sha

            response = self.session.put(
                f"https://api.github.com/repos/{owner}/{repo}/contents/{file_path}",
                headers=headers,
                json=content_data,
            )
            response.raise_for_status()

            pr_data = {
                "head": feature_branch,
                "base": branch,
                "title": commit_message.split("\n")[0],
                "body": pr_description,
            }
            response = self.session.post(
                f"https://api.github.com/repos/{owner}/{repo}/pulls",
                headers=headers,
                json=pr_data,
            )
            response.raise_for_status()
            created_pr = response.json()
            return GitCommitResult(
                success=True,
                branch_name=feature_branch,
                pull_request_url=created_pr.get("html_url", ""),
                pull_request_number=created_pr.get("number", 0),
            )
        except Exception as exc:
            return GitCommitResult(success=False, error_message=str(exc))

    async def add_pull_request_comment(
        self, owner: str, repo: str, pr_number: int, comment: str, token: str
    ) -> bool:
        if not token:
            return False
        headers = {**self.session.headers, "Authorization": f"Bearer {token}"}
        url = f"https://api.github.com/repos/{owner}/{repo}/issues/{pr_number}/comments"
        response = self.session.post(url, headers=headers, json={"body": comment})
        return response.status_code == 201

    async def validate_token(self, token: str) -> TokenValidationResult:
        headers = {**self.session.headers, "Authorization": f"Bearer {token}"}
        try:
            response = self.session.get("https://api.github.com/user", headers=headers)
            if response.status_code == 200:
                return TokenValidationResult(
                    is_valid=True, user_name=response.json().get("login", "")
                )
            return TokenValidationResult(is_valid=False, error_message="Неверный токен")
        except Exception as exc:
            return TokenValidationResult(is_valid=False, error_message=str(exc))
